### PR TITLE
Switch to typescript

### PR DESCRIPTION
**Pull request by harpreetkhalsagtbit**

### CODE REVIEW

File: __test__/index.test.js [hunk 1]
```diff
@@ -1,5 +1,4 @@
-const csc = require('../index.js');
-
+const csc = require('../index.ts');
```
Comment (baywet) [inline __test__/index.test.js]: could be replaced by the import syntax instead `import csc from '../index.ts;'

File: index.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
+import countryList from './lib/country.json';
+import stateList from './lib/state.json';
+import cityList from './lib/city.json';
+
+var country_state_city = {
```
Comment (baywet) [inline index.ts]: export default

File: index.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+function compare(a:any, b:any) {
+	if (a.name < b.name)
+		return -1;
+	if (a.name > b.name)
+		return 1;
+	return 0;
+}
+
+export = country_state_city;
```
Comment (baywet) [inline index.ts]: no need if export default above

File: package.json [hunk 4]
```diff
@@ -2,9 +2,11 @@
   "name": "country-state-city",
   "version": "0.1.2",
   "description": "Library for fetching Country, its States and Cities",
-  "main": "index.js",
+  "main": "dist/index.js",
+  "types": "index.d.ts",
```
Comment (baywet) [inline package.json]: I haven't tried it locally due to lack of time, but are you sure your type def are not in the same folder as your complied code?

File: __test__/index.test.js (hunk 1 above)
Comment (harpreetkhalsagtbit) [reply __test__/index.test.js]: ok, for that I had to convert my .js test files to .ts and run build command again.
Is it what you expected?

File: index.ts (hunk 2 above)
Comment (harpreetkhalsagtbit) [reply index.ts]: I tried this earlier, I was getting some error.
Now it worked the same way.

May be changing require to import fixed it.
Am I right?

File: index.ts (hunk 3 above)
Comment (harpreetkhalsagtbit) [reply index.ts]: yes, working

File: package.json (hunk 4 above)
Comment (harpreetkhalsagtbit) [reply package.json]: forgot to change... it's in `dist/`

**Comment by harpreetkhalsagtbit** [comment]
@baywet 
Please review the updates.

**Comment by harpreetkhalsagtbit** [comment]
when I locally install my package, I can see `dist, lib, __test__ etc` all folders in node_modules, even after I have put `lib and node_modules` in `/npmignore` file. How can I fix it?

File: index.ts (hunk 2 above)
Comment (baywet) [reply index.ts]: I guess this comment https://github.com/harpreetkhalsagtbit/country-state-city/pull/12#discussion_r283086124 means you already found the solution for your question?

File: __test__/index.test.js (hunk 1 above)
Comment (baywet) [reply __test__/index.test.js]: yes that and the import syntax you adopted in your last commit, thanks!

**Comment by baywet** [comment]
the behavior of the npm publish command is that it well take everything located in the current folder.
If you want to exclude anything, it will look at the gitignore file *if no npmignore file exists* otherwise it will only look at the npmignore file.
I suggest you add all the things you want to exclude from your package in your npmignore:
- lib
- index.ts
- jest.config.js
- node_modules
- __test__
- lib/__test__
(on top of what you already have there)
I hope this helps

**Comment by baywet** [comment]
(the test folder name was caught by md, i meant _ _ test _ _ without the spaces)

**Comment by harpreetkhalsagtbit** [comment]
> the behavior of the npm publish command is that it well take everything located in the current folder.
> If you want to exclude anything, it will look at the gitignore file _if no npmignore file exists_ otherwise it will only look at the npmignore file.
> I suggest you add all the things you want to exclude from your package in your npmignore:
> 
> * lib
> * index.ts
> * jest.config.js
> * node_modules
> * **test**
> * lib/**test**
>   (on top of what you already have there)
>   I hope this helps

I did add `lib` and `node_modules` in `.npmignore` file. I haven't published it yet. I was doing local install using
`npm install <path>'

This was adding all the folders to respective node_modules, I wonder it would do same when I run `npm publish`.

Also, please let me know if this P.R is readyToMerge?

**Comment by baywet** [comment]
besides the additional exclusions I was mentioning in my previous comment, it looks good to me. Great work!

**Comment by harpreetkhalsagtbit** [comment]
Thanks @baywet 
It was really a great help. Thanks for all the efforts. Please be in touch for all the future development/improvements/suggestions. I would like to add you as a maintainer of this repository if you want.

**Comment by baywet** [comment]
Thank you! Great team effort! And no, please don't add me as maintainer, I don't have the time to maintain this additional project on a regular basis.

**Comment by harpreetkhalsagtbit** [comment]
@baywet 
after build main index.js returns module as
`exports.default 
instead of 
`module.exports`

Unable to require package and use directly, instead need to call functions using

```js
 var csc = require('country-state-city')
 console.log(csc.default.getAllCountries())
```

it should be 
```js
 var csc = require('country-state-city')
 console.log(csc.getAllCountries())
```

**Comment by baywet** [comment]
hum I think that's because you have other declarations in your files outside of the default one. So it gets transpiled down to a module, containing a default one a a few other ones.
Try to move the two functions inside the export default and see if this changes anything. Not sure it's not going to push a different problem on the functions instead though.

**Comment by harpreetkhalsagtbit** [comment]
export = csc
it got fixed by changing
`export default csc`
to
`export = csc`

But I am not sure whether it will work with `import`

**Comment by harpreetkhalsagtbit** [comment]
Will need to check it thorough, for no I have reverted npm release. Will spend sometime figuring out this issue and will fix it for both require and import if possible

**Comment by harpreetkhalsagtbit** [comment]
@baywet 
I am going with 
```js
export = {
         getCountries() {}
} 

```
I have tested it with `require()` without using '.default` and normally with `import` as well. Its working

**Comment by baywet** [comment]
Just as a follow up, today I upgraded some of my production projects to this latest version.
Works well and is definitively an improvement over the previous version.

However I have an additional suggestion, today methods have an implicit return type or implicit any. If you declared interfaces for country and state, that'd allow consumers to leverage those interfaces in their code and provide even better type check.

```TS
interface ICountry {
    id: string;
    name: string;
    phonecode: string;
    sortname: string;
}
interface IState {
    id: string;
    name: string;
    country_id: string;
}
interface ICity {
   id: string;
   name: string;
   state_id: string;
}
```
And then all you should need to do is type the return of the methods like this
```TS
getCountryById: function (id: string) : ICountry {
		return _findEntry(countryList, id);
	},
```

**Comment by harpreetkhalsagtbit** [comment]
@baywet 
please review
https://github.com/harpreetkhalsagtbit/country-state-city/pull/15

**Pull request merged**